AWS ENI IPAM panic: nil pointer dereference in findSubnetInSameRouteTableWithNodeSubnet

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Version

equal or higher than v1.18.4 and lower than v1.19.0

### What happened?

cilium-operator enters a CrashLoopBackOff and crashes with the following panic:
```
time=2025-11-26T11:09:32.861674215Z level=info msg="Discovered new CiliumNode custom resource" module=operator.operator-controlplane.leader-lifecycle.legacy-cell subsys=ipam-allocator-aws name=ip-172-25-16-12
1.eu-west-3.compute.internal
time=2025-11-26T11:09:32.861779648Z level=info msg="Resolving IP deficit of node" module=operator.operator-controlplane.leader-lifecycle.legacy-cell subsys=ipam-allocator-aws name=ip-172-24-68-235.eu-west-3.c
ompute.internal instanceID=i-0742ee6c7311bb14b selectedInterface=eni-014d2d7b90e734555 selectedPoolID=subnet-00d4c84aada225561 maxIPsToAllocate=1 availableForAllocation=128 emptyInterfaceSlots=2 available=16
used=9 neededIPs=1 remainingInterfaces=3
time=2025-11-26T11:09:32.86178956Z level=info msg="Resolving IP deficit of node" module=operator.operator-controlplane.leader-lifecycle.legacy-cell subsys=ipam-allocator-aws name=ip-172-25-16-121.eu-west-3.co
mpute.internal instanceID=i-0279e919a43ef8ece selectedInterface="" selectedPoolID="" maxIPsToAllocate=8 availableForAllocation=0 emptyInterfaceSlots=2 available=0 used=0 neededIPs=8 remainingInterfaces=3
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x68 pc=0x32915ac]
goroutine 803 [running]:
github.com/cilium/cilium/pkg/aws/eni.(*Node).findSubnetInSameRouteTableWithNodeSubnet(0x40077bdbc0)
    /go/src/github.com/cilium/cilium/pkg/aws/eni/node.go:923 +0x33c
github.com/cilium/cilium/pkg/aws/eni.(*Node).findSuitableSubnet(0x40077bdbc0, {{0x0, 0x0}, {0x4001ed9870, 0xa}, 0x0, 0x0, 0x0, 0x400266f210, {0x0, ...}, ...}, ...)
    /go/src/github.com/cilium/cilium/pkg/aws/eni/node.go:1003 +0xd8
github.com/cilium/cilium/pkg/aws/eni.(*Node).CreateInterface(0x40077bdbc0, {0x50af670, 0x400239f4f0}, 0x4001ce6060, 0x40001a6d70)
    /go/src/github.com/cilium/cilium/pkg/aws/eni/node.go:460 +0x198
github.com/cilium/cilium/pkg/ipam.(*Node).createInterface(0x4001cbbba0, {0x50af670, 0x400239f4f0}, 0x4001ce6060)
    /go/src/github.com/cilium/cilium/pkg/ipam/node.go:573 +0x17c
github.com/cilium/cilium/pkg/ipam.(*Node).handleIPAllocation(0x4001cbbba0?, {0x50af670?, 0x400239f4f0?}, 0x40009301c0?)
    /go/src/github.com/cilium/cilium/pkg/ipam/node.go:956 +0x204
github.com/cilium/cilium/pkg/ipam.(*Node).maintainIPPool(0x4001cbbba0, {0x50af670, 0x400239f4f0})
    /go/src/github.com/cilium/cilium/pkg/ipam/node.go:998 +0x1f8
github.com/cilium/cilium/pkg/ipam.(*Node).MaintainIPPool(0x4001cbbba0, {0x50af670, 0x400239f4f0})
    /go/src/github.com/cilium/cilium/pkg/ipam/node.go:1042 +0xf0
github.com/cilium/cilium/pkg/ipam.(*NodeManager).Upsert.func1({0x40024c1700?, 0x40024c16a4?, 0x7ec2020?})
    /go/src/github.com/cilium/cilium/pkg/ipam/node_manager.go:334 +0x40
github.com/cilium/cilium/pkg/trigger.(*Trigger).waiter(0x400135eaa0)
    /go/src/github.com/cilium/cilium/pkg/trigger/trigger.go:201 +0x470
created by github.com/cilium/cilium/pkg/trigger.NewTrigger in goroutine 650
    /go/src/github.com/cilium/cilium/pkg/trigger/trigger.go:121 +0x1b4
```
The panic occurs systematically right after enabling:
- `eni.subnetIDsFilter` : 
    - We use this setting because we want Cilium to allocate ENIs only from a set of subnets dedicated to Pods. These Pod subnets are different from the subnets used by the EC2 nodes themselves.
Therefore, our intention is that Cilium creates ENIs in the two subnets listed under subnetIDsFilter.
From what I understand, the function _findSubnetInSameRouteTableWithNodeSubnet_ tries to match the route table associated with the node's subnet. **However, the Pod subnets have different routeTable from nodes subnets** (the ones defined in subnetIDsFilter).
    - 100.67.0.0/16 for pods (Cilium ENIs)
    - 172.16.0.0/24 for nodes
- AWS prefix delegation (`awsEnablePrefixDelegation: true`)
- `ipMasqAgent`
- rollout of both the operator and the agents

### How can we reproduce the issue?

1. Start from this initial configuration (working state)
```
cilium:
  eni:
    subnetTagsFilter:
      - karpenter.sh/discovery=NFRA_aws_eks_cluster
    awsReleaseExcessIPs: true

  ipv4NativeRoutingCIDR: 172.16.0.0/12

  k8sServiceHost: <eks-endpoint>
  k8sServicePort: 443
  kubeProxyReplacement: true
  kubeProxyReplacementHealthzBindAddr: "'0.0.0.0:10256'"

  egressMasqueradeInterfaces: ""
  enableIPv4Masquerade: true
  bpf:
    masquerade: true
```
2. Move to the following configuration (crashing state)
```
cilium:
  eni:
    subnetIDsFilter:
      - subnet-0b0ae27e7e58f09f4
      - subnet-00d4c84aada225561

    awsEnablePrefixDelegation: true
    awsReleaseExcessIPs: true

  ipMasqAgent:
    enabled: true
    config:
      nonMasqueradeCIDRs:
        - 100.67.0.0/16
        - 172.24.64.0/21
        - 172.25.16.0/22

  ipv4NativeRoutingCIDR: 172.16.0.0/12

  k8sServiceHost: <eks-endpoint>
  k8sServicePort: 443
  kubeProxyReplacement: true
  kubeProxyReplacementHealthzBindAddr: "'0.0.0.0:10256'"

  egressMasqueradeInterfaces: ""
  enableIPv4Masquerade: true
  bpf:
    masquerade: true
```
3. Common values between our 4 clusters applied during the two previous steps:
```
cilium:
  routingMode: native

  eni:
    enabled: true
    awsReleaseExcessIPs: true

  ipam:
    mode: eni

  encryption:
    enabled: true
    type: wireguard

  egressMasqueradeInterfaces: eth+
  enableIPv4Masquerade: true

  rollOutCiliumPods: true
  policyCIDRMatchMode: nodes
  l7Proxy: false

  operator:
    rollOutPods: true
    unmanagedPodWatcher:
      restart: false
    resources:
      limits:
        cpu: 1
        memory: 1Gi
      requests:
        cpu: 20m
        memory: 80Mi
```
4. Rollout operators / agents

### Cilium Version

v1.18.4

### Kernel Version

Kernel Version:             6.12.55
OS Image:                   Bottlerocket OS 1.51.0 (aws-k8s-1.33)
Operating System:           linux
Architecture:               arm64
Container Runtime Version:  containerd://2.1.5+bottlerocket
Kubelet Version:            v1.33.5-eks-ba24e9c

### Kubernetes Version

Client Version: v1.34.1
Kustomize Version: v5.7.1
Server Version: v1.33.5-eks-3cfe0ce

### Regression

_No response_

### Sysdump

_No response_

### Relevant log output

```shell
time=2025-11-26T11:09:32.861674215Z level=info msg="Discovered new CiliumNode custom resource" module=operator.operator-controlplane.leader-lifecycle.legacy-cell subsys=ipam-allocator-aws name=ip-172-25-16-12
1.eu-west-3.compute.internal
time=2025-11-26T11:09:32.861779648Z level=info msg="Resolving IP deficit of node" module=operator.operator-controlplane.leader-lifecycle.legacy-cell subsys=ipam-allocator-aws name=ip-172-24-68-235.eu-west-3.c
ompute.internal instanceID=i-0742ee6c7311bb14b selectedInterface=eni-014d2d7b90e734555 selectedPoolID=subnet-00d4c84aada225561 maxIPsToAllocate=1 availableForAllocation=128 emptyInterfaceSlots=2 available=16
used=9 neededIPs=1 remainingInterfaces=3
time=2025-11-26T11:09:32.86178956Z level=info msg="Resolving IP deficit of node" module=operator.operator-controlplane.leader-lifecycle.legacy-cell subsys=ipam-allocator-aws name=ip-172-25-16-121.eu-west-3.co
mpute.internal instanceID=i-0279e919a43ef8ece selectedInterface="" selectedPoolID="" maxIPsToAllocate=8 availableForAllocation=0 emptyInterfaceSlots=2 available=0 used=0 neededIPs=8 remainingInterfaces=3
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x68 pc=0x32915ac]
goroutine 803 [running]:
github.com/cilium/cilium/pkg/aws/eni.(*Node).findSubnetInSameRouteTableWithNodeSubnet(0x40077bdbc0)
    /go/src/github.com/cilium/cilium/pkg/aws/eni/node.go:923 +0x33c
github.com/cilium/cilium/pkg/aws/eni.(*Node).findSuitableSubnet(0x40077bdbc0, {{0x0, 0x0}, {0x4001ed9870, 0xa}, 0x0, 0x0, 0x0, 0x400266f210, {0x0, ...}, ...}, ...)
    /go/src/github.com/cilium/cilium/pkg/aws/eni/node.go:1003 +0xd8
github.com/cilium/cilium/pkg/aws/eni.(*Node).CreateInterface(0x40077bdbc0, {0x50af670, 0x400239f4f0}, 0x4001ce6060, 0x40001a6d70)
    /go/src/github.com/cilium/cilium/pkg/aws/eni/node.go:460 +0x198
github.com/cilium/cilium/pkg/ipam.(*Node).createInterface(0x4001cbbba0, {0x50af670, 0x400239f4f0}, 0x4001ce6060)
    /go/src/github.com/cilium/cilium/pkg/ipam/node.go:573 +0x17c
github.com/cilium/cilium/pkg/ipam.(*Node).handleIPAllocation(0x4001cbbba0?, {0x50af670?, 0x400239f4f0?}, 0x40009301c0?)
    /go/src/github.com/cilium/cilium/pkg/ipam/node.go:956 +0x204
github.com/cilium/cilium/pkg/ipam.(*Node).maintainIPPool(0x4001cbbba0, {0x50af670, 0x400239f4f0})
    /go/src/github.com/cilium/cilium/pkg/ipam/node.go:998 +0x1f8
github.com/cilium/cilium/pkg/ipam.(*Node).MaintainIPPool(0x4001cbbba0, {0x50af670, 0x400239f4f0})
    /go/src/github.com/cilium/cilium/pkg/ipam/node.go:1042 +0xf0
github.com/cilium/cilium/pkg/ipam.(*NodeManager).Upsert.func1({0x40024c1700?, 0x40024c16a4?, 0x7ec2020?})
    /go/src/github.com/cilium/cilium/pkg/ipam/node_manager.go:334 +0x40
github.com/cilium/cilium/pkg/trigger.(*Trigger).waiter(0x400135eaa0)
    /go/src/github.com/cilium/cilium/pkg/trigger/trigger.go:201 +0x470
created by github.com/cilium/cilium/pkg/trigger.NewTrigger in goroutine 650
    /go/src/github.com/cilium/cilium/pkg/trigger/trigger.go:121 +0x1b4
```

### Anything else?

A likely nil dereference in:
```
node.go:923
subnet := n.manager.subnets[subnetID]
```
This can be nil if:
- A route table references a subnet ID
- But n.manager.subnets has not yet populated that subnet
(Suggested improvements :add a nil check around the access to `manager.subnets[subnetID]`.)

We run **3 other clusters** with the exact same configuration, except:
- they have far fewer nodes (≈25)
- the larger cluster (~80 nodes) is the only one crashing


Thank you very much for your help and for taking the time to look into this issue :)

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

--- a/pkg/aws/eni/node.go
+++ b/pkg/aws/eni/node.go
@@ -980,6 +980,9 @@ func (n *Node) findSubnetInSameRouteTableWithNodeSubnet() *ipamTypes.Subnet {
 					continue
 				}
 				subnet := n.manager.subnets[subnetID]
+				if subnet == nil {
+					continue
+				}
 				if (bestSubnet == nil || subnet.AvailableAddresses > bestSubnet.AvailableAddresses) && subnet.AvailabilityZone == n.k8sObj.Spec.ENI.AvailabilityZone {
 					bestSubnet = subnet
 				}

--- a/pkg/aws/eni/node_test.go
+++ b/pkg/aws/eni/node_test.go
@@ -115,6 +115,63 @@ func Test_findSubnetInSameRouteTableWithNodeSubnet(t *testing.T) {
 
 }
 
+func Test_findSubnetInSameRouteTableWithNodeSubnet_UntrackedSubnets(t *testing.T) {
+	// This test ensures that the function handles the case where the route table
+	// references subnets that are not tracked by the manager because they were filtered
+	// out by the subnetsFilters parameter
+	routeTableMap := ipamTypes.RouteTableMap{
+		"rt-1": &ipamTypes.RouteTable{
+			ID:               "rt-1",
+			VirtualNetworkID: "vpc-1",
+			Subnets: map[string]struct{}{
+				"subnet-1": {}, // node subnet
+				"subnet-2": {}, // tracked subnet
+				"subnet-3": {}, // untracked subnet (not in manager.subnets)
+				"subnet-4": {}, // another tracked subnet
+			},
+		},
+	}
+
+	node := &Node{
+		k8sObj: &v2.CiliumNode{
+			Spec: v2.NodeSpec{
+				ENI: types.ENISpec{
+					VpcID:            "vpc-1",
+					NodeSubnetID:     "subnet-1",
+					AvailabilityZone: "us-east-1a",
+				},
+			},
+		},
+		manager: &InstancesManager{
+			subnets: map[string]*ipamTypes.Subnet{
+				"subnet-1": {
+					ID:                 "subnet-1",
+					AvailableAddresses: 10,
+					AvailabilityZone:   "us-east-1a",
+				},
+				"subnet-2": {
+					ID:                 "subnet-2",
+					AvailableAddresses: 20,
+					AvailabilityZone:   "us-east-1a",
+				},
+				// subnet-3 is intentionally missing to simulate an untracked subnet
+				"subnet-4": {
+					ID:                 "subnet-4",
+					AvailableAddresses: 30,
+					AvailabilityZone:   "us-east-1a",
+				},
+			},
+			routeTables: routeTableMap,
+		},
+	}
+
+	// This should not panic and should return subnet-4 (highest available addresses)
+	got := node.findSubnetInSameRouteTableWithNodeSubnet()
+	require.NotNil(t, got)
+	require.Equal(t, "subnet-4", got.ID)
+	require.Equal(t, 30, got.AvailableAddresses)
+}
+
 func Test_checkSubnetInSameRouteTableWithNodeSubnet(t *testing.T) {
 	routeTableMap := ipamTypes.RouteTableMap{
 		"rt-1": &ipamTypes.RouteTable{

